keps/sig-apps/961-maxunavailable-for-statefulset/README.md

```diff
@@ -291,89 +720,168 @@ No, the default behavior remains the same.
 
 ### Rollout, Upgrade and Rollback Planning
 
 ###### How can a rollout or rollback fail? Can it impact already running workloads?
 
-A rollout or rollback of this feature can fail if there is a bug which causes the kube-apiserver or 
-the kube-controller-manager to start crashing when the feature flag is enabled.
+<!--
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+-->
 
+It depends, like
 
-Yes, it can impact already running workloads.
+- In a HA cluster, if part of the API servers enabled with this feature gate but the others not, then
+  for already running workloads, they will not be impacted in rolling update because the process is controlled
+  by the statefulset controller.  But for newly created statefulset, their `maxUnavailable` field will
+  be dropped if requested a feature-gate disabled api-server.
 
-If a rolling update is in progress for a StatefulSet, while this feature is being enabled in kube-apiserver
-and kube-controller-manager, the StatefulSet controller can run into corner cases where it will take longer
-for the controller to converge. This will only happen if after enabling the feature, the customer also sets
-maxUnavailable to a number greater than 1,  but the invariants and the logic will ensure that there are never more than
-maxUnavailable pods with the same identity and never more than maxUnavailable being deleted.
+- With this feature enabled and a statefulset running into the rolling-update, then we disabled the feature-gate,
+  then this statefulset will be impacted as it will fall into the default behavior, updating a pod at one time.
 
 ###### What specific metrics should inform a rollback?
-TODO when we reach Beta
+
+<!--
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+-->
+
+- As a normal user, with `maxUnavailable` enabled, and we're rolling update a statefulset, the number of pods brought down should
+equal to the `maxUnavailable`, if not, we should rollback.
+
+- As a administrator(as well for normal user), I can check the new metric `rolling-update-duration-seconds`, if we enabled the feature with
+greater than 1 `maxUnavailable` set for statefulsets, but didn't see the duration fall down, then we may have some problems here.
+This is not precise because rolling-update duration can be impacted by several factors, e.g. pulling down a big, new image, but
+can be some indicators refer to.
```

Comment:
So can we come up with some better metric for it?

We're reporting ReadyReplicas/AvailableReplicas in Status - can we expose a metric if a difference between these and spec.Replicas is more than maxUnavailable as an example?

@soltysh - for thoughts

Reply:
I like the proposed metric a lot more than the rolling time.

Reply:
Sorry, what messages can we get from `ReadyReplicas != AvailableReplicas`, in my understanding, this only relates to another filed `minReadySeconds`.

Reply:
`spec.Replicas - status.readyReplicas` should always be less than or equal to `spec.maxUnavailable`, or this is a bug.
The metric makes little sense I think ...

So for the rolling time, it values not only for the maxUnavailable, it can help to measure how long the rolling update takes, providing data for decisions whether to rolling-update in parallel if possible.  It can be independent of the KEP actually.

Reply:
> spec.Replicas - status.readyReplicas should always be less than or equal to spec.maxUnavailable, or this is a bug.

Which is exactly what we want - we want the metric that shows us if something isn't ok.

FWIW - for me the "rolling-update-duration-seconds" makes very little sense because how should I think about it? How should I compare my 3-replica hello-world with 1000-replica gigantitc java app with 10m startup?
If we want to use for seeing if everything is fine, we need the metric to be somehow denormalizable.

Reply:
> How should I compare my 3-replica hello-world with 1000-replica gigantitc java app with 10m startup

Yes this also makes me doubtable about the metric, ok, then I'll drop this idea and then it will be more simple I think. Thanks.

Reply:
@wojtek-t these information are for the most part exposed in kube-state-metrics today: https://github.com/kubernetes/kube-state-metrics/blob/main/docs/statefulset-metrics.md. We could add maxUnavailable with this effort. Would that be acceptable from a PRR standpoint or is it limited to native metrics?

Reply:
Just for info, kube-state-metrics doesn't follow Kubernetes metrics stability framework just yet, so there are less guarantees towards its metrics than the ones we have in k/k.

But, although that's the case today, it would still be interesting to know if having API related metrics defined in kube-state-metrics would be acceptable from a KEP standpoint since in most cases they are unfit to the Kubernetes codebase.

Reply:
Just to follow-up on it.
We didn't come up with a very concrete conclusion, but the hand-wavy one was:
- for the very simple metrics that effectively monitor a trivial logic on k8s side, we're fine with ksm metrics [like number of objects with a given field etc.]
- for things where the logic of the feature is non-trivial, we actually want a native metric


Now, with this feature we get somewhere on the fence. Let's say we add "maxUnavailable" as a metric to ksm.
IIUC, we can then check the value that I suggested above (replicas-read > maxUnavailable) for a given statefulset purely based on ksm.
Can we easily check how many statefulsets are in a bad state in a given time?


Now thinking further about it, maybe the metric that I would actually want to see is something like
`stateful_set_in_bad_state`  [we obviously need a better name] with a label `reason`.

Now, for this feature we can set `reason=exceededMaxUnavailable` and increase the value of that metric at any time when we process a statefulset and we observe that replicase-ready >maxUnavailable".

That actually seems (a) simple enough to implement (b) generic enough for future extension (c) useful enough as it shows if something goes wrong.

Thoughts?